### ISSUE TITLE
Is it fundamentally possible to communicate information from a particular matrix operation to the outside environment?

### ISSUE THREAD

**Issue by viziionary** [gpujs/gpu.js#318]
**Details:**

Let's say I'm trying to utilize GPU.js for some image recognition, for example detect an in-game character. With GPU.js I can detect edges in real time incredibly fast, like so:

https://www.useloom.com/share/225e05045c534306a53c699f070baf47

```js
var drawFinal = gpu.createKernel(function(inputBuffer, width, height, size, averageColor, averageColorLength) {
    var index = size[0] - (((width[0] * 4) - (this.thread.x * 4)) + (this.thread.y * width[0] * 4));
    var rIndex = index;
    var gIndex = index + 1;
    var bIndex = index + 2;
    var r = inputBuffer[rIndex];
    var g = inputBuffer[gIndex];
    var b = inputBuffer[bIndex];
    this.color(r / 255, g / 255, b / 255);
    for (var x = -1; x < 2; x++) {
        for (var y = -1; y < 2; y++) {
            var indexTarget = size[0] - (((width[0] * 4) - ((this.thread.x + x) * 4)) + ((this.thread.y + y) * width[0] * 4));
            var rIndexTarget = indexTarget;
            var gIndexTarget = indexTarget + 1;
            var bIndexTarget = indexTarget + 2;
            var rTarget = inputBuffer[rIndexTarget];
            var gTarget = inputBuffer[gIndexTarget];
            var bTarget = inputBuffer[bIndexTarget];
            if (Math.abs(rTarget - r) + Math.abs(gTarget - g) + Math.abs(gTarget - g) > 100) {
                //averageColor[0] = ((averageColor[0] * averageColorLength[0]) + r) / (averageColorLength[0] + 2);
                //averageColor[1] = ((averageColor[1] * averageColorLength[0]) + g) / (averageColorLength[0] + 2);
                //averageColor[2] = ((averageColor[2] * averageColorLength[0]) + b) / (averageColorLength[0] + 2);
                this.color(1, 0.3, 0.3);
                //averageColorLength[0] += 1;
            }
        }
    }
})
  .setOutput([width, height])
  .setGraphical(true);
```

You can see in my commented out code I was trying to do the impossible and modify one of the input arrays, but I realize the reason it doesn't work is because GPU operations just don't work that way, they're fundamentally different and don't write to memory outside of each individual operation.

This operation could theoretically let me pinpoint the areas I need to run further processing on, however, I'm not sure how I can share that information with the outside environment in a useful way. I'm sort of crashing into the world of shaders and matrix operations and I realize operations running on the GPU, while really fast, have fundamental limitations. Perhaps this is one of them.

For example, let's say to narrow down the edges detected to only players, first I want to add up the average color of all the edges detected so I can then use that value to avoid the "common" edges throughout the current frame, like trees and mountains and UI features seen here. To do that, I would have to share information with the outside environment of my code. All I can seem to do is create an image with the desired color changes, but that doesn't narrow anything down for further processing because I don't see a way to communicate the key coordinates detected as edges to the outside script. 

**The Question:**

Is what I'm trying to do essentially fundamentally out of reach for GPU operations? Is there any way to end up with an array of only the "edge" pixels?

**Side Note, Unrelated:**

It should be noted that I would not maliciously use or distribute such software (aim assist, potentially) if I figure this out, but I am interested in doing some things for the pure challenge. Like a researcher trying to break into a system to see if it's possible. I happen to think FPS gaming is at risk of these kinds of real time image recognition cheats in the near future, either with neural nets or simple hard coded image recognition like this. I'm not sure if anything can be done about it, since it doesn't involve any actual hacking or memory manipulation. It's been done to some limited extent, but dark environments or distant enemies limits at least basic implementations of the technology significantly. Sufficiently optimized neural networks (such powered by the GPU, like your other project) may be able to overcome those kinds of challenges in real time when combined with shortcuts to narrow down the search space, like with this efficient edge detection system.

**Comment by robertleeplummerjr** [comment]
You can usually do in a GPU operation what you'd be able to do in a CPU operation, you just have to think of them differently.  I think what you are trying to do is _too much in a single algorithm_.  When you run into that issue, simply add another algorithm, or in our case, another kernel/texture.  For example:


```js
var detectEdges = gpu.createKernel(function(inputBuffer, width, size, averageColorLength) {
    var averageColorLength = 0;
    this.color(r / 255, g / 255, b / 255);
    for (var x = -1; x < 2; x++) {
        for (var y = -1; y < 2; y++) {
            var indexTarget = size[0] - (((width[0] * 4) - ((this.thread.x + x) * 4)) + ((this.thread.y + y) * width[0] * 4));
            var rIndexTarget = indexTarget;
            var gIndexTarget = indexTarget + 1;
            var rTarget = inputBuffer[rIndexTarget];
            var gTarget = inputBuffer[gIndexTarget];
            if (Math.abs(rTarget - r) + Math.abs(gTarget - g) + Math.abs(gTarget - g) > 100) {
                averageColorLength += 1;
            }
        }
    }
    return averageColorLength;
})
  .setOutputToTexture(true)
  .setOutput([width, height]);

var drawFinal = gpu.createKernel(function(averageColorLengths, inputBuffer, width, height, size, averageColor, averageColorLength) {
    var index = size[0] - (((width[0] * 4) - (this.thread.x * 4)) + (this.thread.y * width[0] * 4));
    var rIndex = index;
    var gIndex = index + 1;
    var bIndex = index + 2;
    var r = inputBuffer[rIndex];
    var g = inputBuffer[gIndex];
    var b = inputBuffer[bIndex];
    var averageColorLength = averageColorLengths[this.thread.y][this.thread.x];
    if (averageColorLength > 0) {
        this.color(
            ((averageColor[0] * averageColorLength) + r) / (averageColorLength + 2),
            ((averageColor[1] * averageColorLength) + g) / (averageColorLength + 2),
            ((averageColor[2] * averageColorLength) + b) / (averageColorLength + 2)
        );
    } else {
        this.color(r / 255, g / 255, b / 255);
    }
})
  .setGraphical(true)
  .setOutput([width, height]);
```


You are simply breaking up the working into two algorithms, then just, send the output of the first into the second for visualization.  I've not tested the above to see if it totally works, but the math and logic is simple enough, as well as past experience, that this is simply a task of breaking the work down into its simpler components.

**Comment by robertleeplummerjr** [comment]
So to your question, if I properly understand it:
>Is it fundamentally possible to communicate information from a particular matrix operation to the outside environment?

I'd say, "Yes!  You just have to understand how the gpu works and take advantage of it, don't fight it."

**Comment by robertleeplummerjr** [comment]
I would really like to see how you feed the video into gpu.js, if that is actually being done.  Would you mind sharing?  We could post it in the docs for reuse.

**Comment by viziionary** [comment]
Thanks for the explanation. And yeah my code was:

```
window.onload = function() {
    var video = document.getElementById('video');
    var flag = 0;
    var averageColor = [0, 0, 0];
    var averageColorLength = 0;
    video.src = 'test.mp4';
    video.load();
    video.addEventListener('loadeddata', function() {
        var edges;
        var frameBuffer;
        var mouseX = 0;
        var mouseY = 0;
        var width = video.videoWidth;
        var height = video.videoHeight;
        var canvasSource = document.createElement('canvas');
        canvasSource.width = width;
        canvasSource.height = height;
        document.body.appendChild(canvasSource);
        var context = canvasSource.getContext('2d');
        var canvasTarget = document.createElement('canvas');
        canvasTarget.width = width;
        canvasTarget.height = height;
        document.body.appendChild(canvasTarget);
        var gl = canvasTarget.getContext('webgl');
        var gpu = new GPU({
            canvas: canvasTarget,
            gl: gl
        });
        var size = width * height * 4;
        var drawFinal = gpu.createKernel(function(inputBuffer, width, height, size, averageColor, averageColorLength) {
            var index = size[0] - (((width[0] * 4) - (this.thread.x * 4)) + (this.thread.y * width[0] * 4));
            var rIndex = index;
            var gIndex = index + 1;
            var bIndex = index + 2;
            var r = inputBuffer[rIndex];
            var g = inputBuffer[gIndex];
            var b = inputBuffer[bIndex];
            this.color(r / 255, g / 255, b / 255);
            for (var x = -1; x < 2; x++) {
                for (var y = -1; y < 2; y++) {
                    var indexTarget = size[0] - (((width[0] * 4) - ((this.thread.x + x) * 4)) + ((this.thread.y + y) * width[0] * 4));
                    var rIndexTarget = indexTarget;
                    var gIndexTarget = indexTarget + 1;
                    var bIndexTarget = indexTarget + 2;
                    var rTarget = inputBuffer[rIndexTarget];
                    var gTarget = inputBuffer[gIndexTarget];
                    var bTarget = inputBuffer[bIndexTarget];
                    if (Math.abs(rTarget - r) + Math.abs(gTarget - g) + Math.abs(gTarget - g) > 100) {
                        //averageColor[0] = ((averageColor[0] * averageColorLength[0]) + r) / (averageColorLength[0] + 2);
                        //averageColor[1] = ((averageColor[1] * averageColorLength[0]) + g) / (averageColorLength[0] + 2);
                        //averageColor[2] = ((averageColor[2] * averageColorLength[0]) + b) / (averageColorLength[0] + 2);
                        this.color(1, 0.3, 0.3);
                        //averageColorLength[0] += 1;
                    }
                }
            }
        }).setOutput([width, height]).setGraphical(true);

        function detect() {
            context.drawImage(video, 0, 0, width, height);
            frameBuffer = context.getImageData(0, 0, width, height).data;
            drawFinal(frameBuffer, [width], [height], [size], averageColor, [averageColorLength]);
            if (flag < 10) {
                console.log(averageColor)
            }
            flag++;
        }

        function run() {
            requestAnimationFrame(run);
            detect();
        }

        video.play();
        run();
    }, false);
}
```

**Comment by robertleeplummerjr** [comment]
I'm super surprised you get such good results.  We are actually making it somewhat difficult, because the video really already exists on the GPU, so, in affect, you have to tell the gpu to send the memory of the video to the cpu, to send it back to the gpu (😛).  If we supported video input (via https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/texImage2D, which we are doing already for images...  added recently) you'd get ridiculously good performance, and could feed the video element directly into your edge detector, and potentially a video driven neural network.


The suggestion above, did it help you?  Did it answer your questions for this issue?

**Comment by viziionary** [comment]
It did put me on the right track I think, to understanding how to proceed. You can close this, I understand things.

Side Note:

It would be really neat to try object recognition neural nets with this edge detection technique! You can input like ~5% of the pixels (edges) in the bitmap to the input layer rather than feeding all the pixels and the system should be able to learn to recognize objects based on their edges alone, since it does separate facial features as you can see in the video example. With your GPU driven system we could possibly do efficient real time object recognition with that technique whereas projects that I know of who are doing object recognition are using entire GPU farms and its not even close to real time.

**Comment by robertleeplummerjr** [comment]
>It would be really neat

Neat, yes, but also very helpful.  Think of self driving cars architected with simple javascript.

>With your GPU driven system

Brain.js v2, which I'm working on now (specifically I'm working on convolutions, and then I've got some work to do on recurrent nets, and then release time!) will essentially be all gpu, minus the `NeuralNetwork` class (however, there is a `NeuralNetworkGPU` class, which we used as a proof of concept).  All we'd need to do is add video input, and the rest can easy be worked out on the neural network.  The only thing missing is more maintainers, more help.

These are tools _I want to use_ in node.  I do not want to have tons of different languages and server farms when we aren't taking full advantage of the hardware on a single device.  Gpu.js aims at solving that.

As for helping here, I just opened https://github.com/gpujs/gpu.js/issues/320.  Either here or there (gpu.js or brain.js) a simple proof of concept can change the world.

I value your feedback.  Ty for pushing this library and seeing what it can do.
